examples/restricted-assets/test/asset-txfer-test.js

```diff
@@ -0,0 +1,332 @@
+  const clearProgram = getProgram('poi-clear.teal');
+
+  this.beforeEach(async function () {
+    alice = new StoreAccount(minBalance, { addr: aliceAddr, sk: new Uint8Array(0) });
+    bob = new StoreAccount(minBalance, { addr: bobAddr, sk: new Uint8Array(0) });
+    runtime = new Runtime([master, alice, bob]);
+
+    creationFlags = {
+      sender: alice.account,
+      localInts: 1,
+      localBytes: 0,
+      globalInts: 2,
+      globalBytes: 1
+    };
+
+    /* Create asset + optIn to asset */
+    assetId = runtime.createAsset('gold', { creator: { ...alice.account, name: 'alice' } });
+    assetDef = runtime.getAssetDef(assetId);
+
+    assert.equal(assetDef.creator, alice.address);
+    assert.equal(assetDef['default-frozen'], true);
+    assert.equal(assetDef.total, 1000000);
+    assert.equal(assetDef['unit-name'], 'GLD');
+    assert.equal(assetDef.manager, alice.address);
+    assert.equal(assetDef.clawback, alice.address);
+
+    runtime.optIntoASA(assetId, bob.address, {});
+    const aliceAssetHolding = runtime.getAssetHolding(assetId, aliceAddr);
+    const bobAssetHolding = runtime.getAssetHolding(assetId, bobAddr);
+    assert.isDefined(aliceAssetHolding);
+    assert.isDefined(bobAssetHolding);
+
+    /* Create application + optIn to app */
+    const creationArgs = [
+      `int:${assetId}`,
+      'int:2' // set min user level(2) for asset transfer ("Accred-level")
+    ];
+
+    applicationId = runtime.addApp(
+      { ...creationFlags, appArgs: creationArgs }, {}, approvalProgram, clearProgram);
+
+    const app = alice.getApp(applicationId);
+    const alicePk = addressToPk(alice.address);
+
+    // verify global state after app creation
+    assert.isDefined(app);
+    assert.deepEqual(getGlobal('Creator'), alicePk);
+    assert.deepEqual(getGlobal('AssetID'), BigInt(assetId));
+    assert.deepEqual(getGlobal('AssetLevel'), 2n); // we are setting level 2 in examples
+  });
+
+  const getGlobal = (key) => runtime.getGlobalState(applicationId, key);
+  const getEscrowProg = (assetId, appId) =>
+    getProgram('clawback-escrow.py', { ASSET_ID: assetId, APP_ID: appId });
+
+  // Update account state
+  function syncAccounts () {
+    alice = runtime.getAccount(alice.address);
+    bob = runtime.getAccount(bob.address);
+    if (escrow) { escrow = runtime.getAccount(escrow.address); }
+  }
+
+  it('should transfer 1000 Assets from Alice to Bob according to custom logic', () => {
+    /**
+     * This test demonstrates how to transfer assets from account A to B using custom logic
+     * based on a smart contract. Asset is actually transferred by the clawback address (an escrow
+     * account in this case). Following operations are performed
+     * - Create the asset + optIn
+     * - Create the application + optIn
+     * - Setup Escrow Account
+     * - Update Asset clawback + lock manager and freeze address
+     * - Set Accred-Level
+     * - Transfer Asset from Alice -> Bob
```

Comment:
Please modify this line to revoke assets

Reply:
Will add revoking instead of modifying. From the top level it's just a transfer of asset. Revoking via clawback escrow is internal logic

Reply:
ok